### PR TITLE
Add API for configuring minimum connections for HTTP/2 connection pool

### LINKED ISSUE #1808: HTTP2 - Round-robining streams across a minimum setting of connections
We leverage L4 load balancers for distributing load, resiliency needs, etc.  We are thinking 2 additions (likely opt-in) to the HTTP2 pool could help us in this regard:

1. The HTTP2 pool maintains a minimum number of connections before opening additional streams.
2. As it opens additional streams, it round-robins between the connections in the HTTP2 pool.  A fancier technique where it chooses the conn with the least number of open streams is likely better, but it's unclear yet if that is critical.

Some of the above was initially discussed [here](https://gitter.im/reactor/reactor-netty?at=6127c20913ac9b6b83969047).

### PR DESCRIPTION
**Pull request by violetagg**
Fixes #1808

### CODE REVIEW

File: reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java [hunk 1]
```diff
@@ -80,6 +80,10 @@
  *     <li>{@link PoolMetrics#idleSize()} always returns {@code 0}.</li>
  * </ul>
  * <p>
+ * If minimum connections is specified, the cached connections with active streams will be kept at that minimum
+ * (can be the best effort). However, if the cached connections have reached max concurrent streams,
+ * then new connections will be allocated up to the maximum connections limit.
```
Comment (crankydillo) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java:85]: Hmm, so the code now will do this, right:
For any new HTTP/2 request, if the number of socket connections is below the minimum number of connections, a new socket connection will get created and the HTTP/2 stream will get created from that channel, right?

I don't really get that from the javadoc, but that may just be me.

What does the best effort part mean?  I think I'm missing that part in the code and want to make sure I'm not misunderstanding something?

**Review by crankydillo — COMMENTED** [review]
I understand a decent bit, and it looks good to me.  Have a question about the doc though.

File: reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java (hunk 1 above)
Comment (violetagg) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java:85]: > Hmm, so the code now will do this, right: For any new HTTP/2 request, if the number of socket connections is below the minimum number of connections, a new socket connection will get created and the HTTP/2 stream will get created on that new socket connection, right?

yes

> I don't really get that from the javadoc, but that may just be me.
> 
> What does the best effort part mean? I think I'm missing that part in the code and want to make sure I'm not misunderstanding something?

As we do not have a strong synchronisation between when we return the connection and we acquire the connection we may not be able to stay exactly on the min number that is specified.

**Comment by violetagg** [comment]
I need to think about the proper place for this configuration. Most probably on the level of `ConnectionProvider` as we can have connection pools per remote host.

**Comment by samueldlightfoot** [comment]
Thanks for this - we are very keen to consume this feature to aid our client-side load balancing too!

**Comment by violetagg** [comment]
This is superseded by #2257

A new API `Http2AllocationStrategy` is exposed so that minimum connections and maximum concurrent streams can be configured.

**Pull request closed without merging**